### ISSUE TITLE
Wrong derivative backpropagating through Cholesky factorization?

### ISSUE THREAD

**Issue by jacobrgardner** [pytorch/pytorch#18825]
## 🐛 Bug
I am getting different derivatives when I compute `g(A)` via `A -> g(A)` versus `A -> LL' -> g(LL')`. The derivatives are off in a predictable way, and it's easy to correct (~2 lines of code), leading me to believe this is a bug.

## To Reproduce

```python
import torch

mat = torch.randn(4, 4, dtype=torch.float64)
mat = (mat @ mat.transpose(-1, -2)).div_(5).add_(torch.eye(4, dtype=torch.float64))
mat = mat.detach().clone().requires_grad_(True)
mat_clone = mat.detach().clone().requires_grad_(True)

# Way 1
inv_mat1 = mat_clone.inverse()  # A^{-1} = A^{-1}

# Way 2
chol_mat = mat.cholesky()
chol_inv_mat = chol_mat.inverse().transpose(-2, -1)
inv_mat2 = chol_inv_mat @ chol_inv_mat.transpose(-2, -1)  # A^{-1} = L^{-T}L^{-1}

# True
print('Are these both A^{-1}?', bool(torch.norm(inv_mat1 - inv_mat2) < 1e-8))

inv_mat1.trace().backward()
inv_mat2.trace().backward()

print('Way 1\n', mat_clone.grad)
print('Way 2\n', mat.grad)  # :-(

corrected_deriv = mat.grad.clone() / 2
corrected_deriv = corrected_deriv.tril() + corrected_deriv.tril().t()
print('Corrected derivative\n', corrected_deriv)  # Simple correction to derivative works.

# True
print('Is the corrected derivative correct?', bool(torch.norm(corrected_deriv - mat_clone.grad) < 1e-8))
```

## Expected behavior
Calling `inv_mat2.trace().backward()` should produce `corrected_deriv` in `mat.grad`, not what is currently going there.

**Comment by soumith** [comment]
@vishwakftw

**Comment by vishwakftw** [comment]
Thank you for reporting this. I will take a look ASAP, and get back to you.

**Comment by vishwakftw** [comment]
I have taken a look at the code for the derivative of the Cholesky decomposition. It seems that the derivative is intentionally left to be a lower / upper triangular matrix depending on the `upper` argument in `torch.cholesky` based on equation 10 in [this paper](https://arxiv.org/pdf/1602.07527.pdf). This is why the derivative that you see is triangular matrix.

@colesbury will probably have a better idea of why it was left this way.

**Comment by gpleiss** [comment]
Upon further investigation, I think that the derivative is correct, but misleading.

As it is implemented, `mat.grad` produced by `torch.cholesky` is lower triangular. This is because `mat.grad[i, j]` is assigned the gradient for `mat[i, j] = mat[j, i]`.

Take the PD matrix:
```python
mat = torch.tensor([
    [3, 1, 0.5, 0],
    [1, 3, -1,  1],
    [0.5, -1, 4, 2],
    [0, 1, 2, 4]
])
```

If you construct `mat` in the following way:
```python
off_diag = torch.tensor([
    [0.0000,  0.0000,  0.0000,  0.0000],
    [ 1.0000,  0.0000,  0.0000,  0.0000],
    [ 0.5000, -1.0000,  0.0000,  0.0000],
    [ 0.0000,  1.0000,  2.0000,  0.0000]
], requires_grad=True)
diag = torch.tensor([3, 3, 4, 4], requires_grad=True)
mat = off_diag + off_diag.t() + diag.diag()
```

then the gradients of `off_diag` and `diag` be the same for Cholesky solves as well as direct inversion solves.

**Comment by jacobrgardner** [comment]
In my opinion, the derivative for `A` should be spread to fill the entire matrix (e.g., like the corrected derivative that I give in my example code). Because `Tr(A^{-1})` has a unique derivative w.r.t. `A`, computing it via `A.inverse().trace().backward()` or via `Tr(L^{-T}L^{-1})` should produce the same derivative for `A` because they both compute the same scalar function `f(A)`.

**Comment by vishwakftw** [comment]
I too agree with @jacobrgardner. There shouldn't be an ambiguity while computing gradients for function that use / don't use the Cholesky decomposition. I'm labeling this as `needs discussion` for the time being.

**Comment by ezyang** [comment]
Hi @vishwakftw, just to double check; what question exactly do we need to resolve, before we can fix this issue?

**Comment by vishwakftw** [comment]
The question is whether to use a full matrix gradient or a half (triangular) matrix gradient. If we use the latter, this issue will be unresolved for reasons specified in [this comment](https://github.com/pytorch/pytorch/issues/18825#issuecomment-479998806).

I also checked with TensorFlow, and they seem to provide a full matrix gradient instead of a triangular matrix gradient.

**Comment by ezyang** [comment]
cc @fmassa 

Also cc @SsnL, ISTR you dealt with autograd functions with symmetry requirements when you worked on FFT. Maybe you have an idea?

**Comment by SsnL** [comment]
@vishwakftw So IIUC this issue, this is due to that `cholesky` is not a proper function in `R^{MxM}`, but only in the subspace of symmetric matrices. 

In such cases we must acknowledge that gradients are not unique (or not properly defined). (e.g., think `f(v) = sum(v) = v[0]*2 = v[1]*2 = 1.1*v[0]+0.9*v[1]` only defined for `{(x, x) : x in R}`.)

Therefore, no matter which way of gradient computation we we decide, there will be inconsistencies with gradients from another seeming identical function (if one only look at forward outputs).

That said, I vote for returning a full symmetric gradient here because it 

  + gives more stable update in terms of gradient-based optimization;
  + in SGD-like algorithms, ensures that the updated matrix is still symmetrical;
  + is the intuitively expected thing to do.

**Comment by ezyang** [comment]
SGTM.

**Comment by vishwakftw** [comment]
Closed by https://github.com/pytorch/pytorch/commit/3403cb857bc2e1bdfb26665688f077ee1b05e811.

**Comment by calincru** [comment]
> In such cases we must acknowledge that gradients are not unique (or not properly defined). (e.g., think `f(v) = sum(v) = v[0]*2 = v[1]*2 = 1.1*v[0]+0.9*v[1]` only defined for `{(x, x) : x in R}`.)
> 
> Therefore, no matter which way of gradient computation we we decide, there will be inconsistencies with gradients from another seemingly identical function (if one only look at forward outputs).
> 
> That said, I vote for returning a full symmetric gradient here because it
> 
> * gives more stable update in terms of gradient-based optimization;
> * in SGD-like algorithms, ensures that the updated matrix is still symmetrical;
> * is the intuitively expected thing to do.

I agree with the conclusion but I disagree with the motivation. In particular, the first statement is wrong. From a mathematical perspective, the function `f(x1, x2, ..., xn) = x1 + x2 + ... + xn` defined on the line `{x in R^n | x1 = x2 = ... = xn}` has a well defined gradient.  Just because in this case we can think of many functions that are equivalent, like you exemplified, doesn't mean that there is an ambiguity in terms of which one we should consider when computing the gradients. Indeed, I want the gradients for **my** function, not those of an equivalent re-parametrization.

**Comment by SsnL** [comment]
@calincru Well, the gradient is really only meaningful if you can make infinitesimal perturbations near the point you are taking gradient. If you are going for strict mathematical perspective, neither your nor my example make sense, as they just don't have support outside that linear subspace. 

In that subspace, there is no such thing as **your** function or not your function, as long as the input-output mapping stays the same. In your argument about being your function or not, only makes sense if you extend the function beyond the subspace.

What we are doing here essentially is very similar. It is just "making up" a function that supports such perturbations, and is equivalent in the subspace. What I was saying above, is just that there are different ways to do this.

**Comment by calincru** [comment]
> @calincru Well, the gradient is really only meaningful if you can make infinitesimal perturbations near the point you are taking gradient. If you are going for strict mathematical perspective, neither your nor my example make sense, as they just don't have support outside that linear subspace.
> 

I'm not sure that's true, though.  The gradient is defined as the unique vector `x` for which the inner product `<x,v>` is equal to the directional derivative at `x` along `v`, for all `v`.  With this definition I don't see any problem with defining gradients for the previous examples. The only caveat is to make it clear what "for all `v`" means: in this case, that's the linear subspace itself.  In essence, your *perturbations*-based definition takes an extrinsic view of the space on which the function is defined: it assumes (or uses the knowledge) that the domain is a subspace of an Euclidean space.  BTW, this way of defining the gradient is exactly how they are generalized in Riemannian geometry.

> In that subspace, there is no such thing as **your** function or not your function, as long as the input-output mapping stays the same. In your argument about being your function or not, only makes sense if you extend the function beyond the subspace.
>

No, **my** function is *different* than other re-parametrizations specifically because it's not just the input-output equivalence that matters.

**Comment by EFitzke** [comment]
This was closed in 2019 after [vishwakftw's](https://github.com/vishwakftw) commit, however, when I run [jacobrgardner's](https://github.com/jacobrgardner) example from above in pytorch 2.0 I still get the incorrect derivative. 

> Closed by [3403cb8](https://github.com/pytorch/pytorch/commit/3403cb857bc2e1bdfb26665688f077ee1b05e811).

**Comment by SebastianAment** [comment]
@EFitzke Which gradients are you comparing? When I execute the example and check `bool(torch.norm(mat.grad - mat_clone.grad) < 1e-8)`, it is true on the main branch. However, `bool(torch.norm(corrected_deriv - mat_clone.grad) < 1e-8)` isn't true any more since the correction was merged.
